### ISSUE TITLE
Headers are missing when consuming a tombstone message from the topic

### ISSUE THREAD

**Issue by PiotrMichalowski96** [spring-cloud/spring-cloud-function#1060]
**Describe the bug**
Issue occurred after migrating from Spring Cloud 2021.0.3 to 2022.0.3. When it consumes a message with a null payload from the topic, the whole message becomes null, causing the headers (including the message key) to be lost.

**Sample**
I am attaching a sample code to reproduce issue. I ran it with "spring-cloud-starter-stream-kafka" in version 2022.0.3:
```
@SpringBootTest
@TestPropertySource(properties = {
    "spring.cloud.stream.kafka.binder.brokers=localhost:9092",
    "spring.cloud.stream.bindings.printer-out-0.destination=test",
    "spring.cloud.stream.kafka.bindings.printer-out-0.producer.configuration.key.serializer=org.apache.kafka.common.serialization.StringSerializer",
    "spring.cloud.stream.kafka.bindings.printer-out-0.producer.configuration.value.serializer=org.springframework.kafka.support.serializer.JsonSerializer",
    "spring.cloud.stream.bindings.printer-in-0.destination=test",
    "spring.cloud.stream.kafka.bindings.printer-in-0.consumer.configuration.key.deserializer=org.apache.kafka.common.serialization.StringDeserializer",
    "spring.cloud.stream.kafka.bindings.printer-in-0.consumer.configuration.value.deserializer=org.springframework.kafka.support.serializer.JsonDeserializer"
})
class KafkaNullConsumerApplicationTests {

  @Autowired
  private StreamBridge streamBridge;

  @Autowired
  private KafkaNullConsumer kafkaNullConsumer;

  @Test
  void shouldConsume() throws InterruptedException {
    final var id = UUID.randomUUID().toString();

    Message<KafkaNull> tombstoneEvent = MessageBuilder
        .withPayload(KafkaNull.INSTANCE)
        .setHeader(KafkaHeaders.KEY, id)
        .build();

    Thread.sleep(5000);
    streamBridge.send("printer-out-0", tombstoneEvent);

    boolean messageConsumed = kafkaNullConsumer.latch().await(2, TimeUnit.SECONDS);
    assertThat(messageConsumed).isTrue();
  }
}

@Slf4j
@Component
class KafkaNullConsumer {

  private final CountDownLatch latch = new CountDownLatch(2);

  @Bean
  public Consumer<Message<KafkaNull>> printer() {
    return tombstoneMessage -> {
      latch.countDown();
      if (tombstoneMessage == null) {
        log.warn("Cannot find message headers cause message is null");
      } else {
        latch.countDown();
        tombstoneMessage.getHeaders();
        log.info("Headers: {}", tombstoneMessage.getHeaders());
      }
    };
  }

  public CountDownLatch latch() {
    return latch;
  }
}
```

**Comment by PiotrMichalowski96** [comment]
I investigated the issue a bit, and it seems that the reason could be in this place in the code:
![missing_headers_issue](https://github.com/spring-cloud/spring-cloud-function/assets/57149032/a7e8627b-8e43-41a7-ad49-c1115832a509)

**Comment by jamfor352** [comment]
Are there any updates on this?

Being able to handle tombstone messages while preserving headers is pretty crucial - if I'm going to be consuming a tombstone, obviously I need to know about the context behind the deletion so I can act upon it.

Please can the developers not return `.accept(null)`? This whole if statement is unnecessary, simply always return `.accept(extractedValue)` - really unsure why this code to remove information even exists.

**Comment by jamfor352** [comment]
@PiotrMichalowski96 seems like there aren't any updates by the Spring Cloud Function team at the moment, but if you're still having issues with this, I can suggest a workaround which works for me. It's annoying, but works:

Instead of

```
@Bean
  public Consumer<Message<KafkaNull>> printer() {
    return tombstoneMessage -> {
      latch.countDown();
      if (tombstoneMessage == null) {
        log.warn("Cannot find message headers cause message is null");
      } else {
        latch.countDown();
        tombstoneMessage.getHeaders();
        log.info("Headers: {}", tombstoneMessage.getHeaders());
      }
    };
  }
```

Simply use a `Function<Message<KafkaNull>, Void>` bean and make sure to return `null`:

```
@Bean
  public Function<Message<KafkaNull>, Void> printer() {
    return tombstoneMessage -> {
      latch.countDown();
      if (tombstoneMessage == null) {
        log.warn("Cannot find message headers cause message is null");
      } else {
        latch.countDown();
        tombstoneMessage.getHeaders();
        log.info("Headers: {}", tombstoneMessage.getHeaders());
      }
      return null;
    };
  }
 ```
and in your YAML, create a 
`printer-out-0:` binding with the destination set to the same as the input binding for `printer-in-0` (this prevents unnecessary topics being created).

Basically, looks like this bug (although from the code it seems intentional?) only is exhibited on Consumers, whereas Functions are left intact. This solution creates a Function which basically acts as a consumer and then returns null to make sure nothing is actually published.

**Comment by olegz** [comment]
Sorry, we were busy with our annual conference, back now and will look shortly

**Comment by sobychacko** [comment]
@PiotrMichalowski96 @jamfor352 This was indeed a regression issue. It is now fixed upstream in `main` (`4.1.x`) and back-ported to `4.0.x`. Sorry for the delay in fixing and thanks for pointing out this issue.
